cardano-db/src/Cardano/Db/AlterTable.hs

```diff
@@ -0,0 +1,95 @@
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# OPTIONS_GHC -Wno-unused-local-binds #-}
+
+module Cardano.Db.AlterTable (
+  AlterTable(..),
+  DbAlterTableException(..),
+  alterTable,
+) where
+
+import Control.Exception.Lifted (Exception, handle, throwIO)
+import Control.Monad.IO.Class (MonadIO, liftIO)
+import Control.Monad.Trans.Control (MonadBaseControl)
+import Control.Monad.Trans.Reader (ReaderT)
+import qualified Data.Text as T
+import Database.Persist.Postgresql (ConstraintNameDB (..), EntityNameDB (..), FieldNameDB (..), SqlBackend, rawExecute, EntityDef, getEntityFields, fieldDB)
+import Database.PostgreSQL.Simple (SqlError (..), ExecStatus (..))
+import Database.Persist.EntityDef.Internal (entityDB)
+
+-- The ability to `ALTER TABLE` currently dealing with `CONSTRAINT` but can be extended
+data AlterTable
+  = AddUniqueConstraint ConstraintNameDB [FieldNameDB]
+  | DropUniqueConstraint ConstraintNameDB
+  deriving (Show)
+
+data DbAlterTableException
+  = DbAlterTableException String SqlError
+  deriving (Show)
+
+instance Exception DbAlterTableException
+
+alterTable ::
+  forall m.
+  ( MonadBaseControl IO m
+  , MonadIO m
+  ) =>
+  EntityDef ->
+  AlterTable ->
+  ReaderT SqlBackend m ()
+alterTable entity (AddUniqueConstraint cname cols) = do
+  -- Check that input fields are indeed present
+  if checkAllFieldsValid entity cols
+    then handle alterTableExceptHandler (rawExecute query [])
+    else liftIO $ throwIO (DbAlterTableException "Constraint field does not exist" sqlError)
+  where
+    query :: T.Text
+    query =
+      T.concat
+        [ "ALTER TABLE "
+        , unEntityNameDB (entityDB entity)
+        , " ADD CONSTRAINT IF NOT EXISTS "
```

Comment:
Is `ADD CONSTRAINT IF NOT EXISTS` a valid close? I can't seem to find it in https://www.postgresql.org/docs/current/sql-altertable.html

Reply:
I'm sure I saw it somewhere when I was researching `ADD CONSTRAINT...` I thought above is what you meant when you wrote this in the issue:

> Also before creating the Unique Key it's necessary to first check if it has already been created before. Postgres queries usually have a IF NOT EXISTS clause for this. Alternatively a separate query needs to check it.

Have I got it in the wrong order?

Reply:
If they cannot be in the same query, then it should be two separate queries.

Reply:
might be able to have it in the same query, just unsure if the way/order I've written it is right. Will search again to see how `IF NOT EXISTS` should be used in this scenario.

Reply:
You're right it doesn't exists only `IF EXISTS`.